packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,200 @@
+		legacyConfig: false,
+		modernConfig: true,
+	},
+	flushMode: {
+		// Note: 1.x clients are compatible with TurnBased flushing, but here we elect to remain on Immediate flush mode
+		// as a work-around for inability to send batches larger than 1Mb. Immediate flushing keeps batches smaller as
+		// fewer messages will be included per flush.
+		minVersionForModernConfig: "2.0.0",
+		legacyConfig: FlushMode.Immediate,
+		modernConfig: FlushMode.TurnBased,
+	},
+	gcOptions: {
+		// Explicitly disable running Sweep in compat mode "2". Although sweep is supported in 2.x, it is disabled by default.
+		// This setting explicitly disables it to be extra safe.
+		minVersionForModernConfig: "3.0.0",
+		legacyConfig: {},
+		modernConfig: { enableGCSweep: true },
+	},
+};
+
+/**
+ * Returns the default version-dependent configs for a given compatibility mode.
+ */
+export function getConfigsForCompatMode(
+	compatibilityMode: Required<IContainerRuntimeOptionsInternal>["compatibilityMode"],
+): IContainerRuntimeOptionsVersionDependent {
+	const defaultConfigs = {
+		gcOptions: {},
+		flushMode: FlushMode.TurnBased,
+		compressionOptions: enabledCompressionConfig,
+		enableRuntimeIdCompressor: undefined as unknown as "on" | "delayed",
+		enableGroupedBatching: true,
+		explicitSchemaControl: false,
+	};
+
+	// TODO: Remove this block after 3.0 is released.
+	// Note: we compare `compatibilityMode` with the exact string "pre-3.0-default" in case we modify `defaultCompatibilityMode` in the future,
+	// but forget to remove this block.
+	if (compatibilityMode === "pre-3.0-default") {
+		return defaultConfigs;
+	}
+
+	for (const key of Object.keys(versionDependentOptionConfigMap)) {
+		const config =
+			versionDependentOptionConfigMap[key as keyof IContainerRuntimeOptionsVersionDependent];
+		// If the compatibility mode is greater than or equal to the minimum version
+		// required for this option, use the "modern" config value, otherwise use the "legacy" config value
+		const isModernConfig =
+			config.minVersionForModernConfig === undefined
+				? false // If the minVersionForModernConfig is undefined, we always use the legacy config
+				: semver.gte(compatibilityMode, config.minVersionForModernConfig);
+		defaultConfigs[key] = isModernConfig ? config.modernConfig : config.legacyConfig;
+	}
+	return defaultConfigs;
+}
+
+/**
+ * Checks if the compatibility mode is valid.
+ * A valid compatibility mode is a string that is a valid semver version and is less than or equal to the current package version.
+ */
+export function isValidCompatMode(
+	compatibilityMode: Required<IContainerRuntimeOptionsInternal["compatibilityMode"]>,
+): boolean {
+	return (
+		// TODO: We can remove the first condition after 3.0 is released and the defaultCompatibilityMode is set to "2.0.0".
+		compatibilityMode === defaultCompatibilityMode ||
+		(compatibilityMode !== undefined &&
+			semver.valid(compatibilityMode) !== null &&
+			semver.lte(compatibilityMode, pkgVersion))
```

Comment:
How much have you considered how this interacts with prerelease tags? I think there are some interesting implications based on [the spec](https://github.com/npm/node-semver?tab=readme-ov-file#prerelease-tags)...

Reply:
ok, I guess the resolution spec there only applies when npm is trying to select versions satisfying a range but not for lte/gte. Cool.